crates/sui/src/client_commands.rs

```diff
@@ -3098,3 +3107,66 @@ async fn check_protocol_version_and_warn(client: &SuiClient) -> Result<(), anyho
 
     Ok(())
 }
+
+/// Fetch move packages based on the provided package IDs.
+async fn fetch_move_packages(
+    client: &SuiClient,
+    package_ids: &[ObjectID],
+) -> Result<Vec<MovePackage>, anyhow::Error> {
+    let objects = client
+        .read_api()
+        .multi_get_object_with_options(package_ids.to_vec(), SuiObjectDataOptions::bcs_lossless())
+        .await
+        .map_err(|e| anyhow!("{e}"))?;
+
+    objects
+        .into_iter()
+        .map(|o| {
+            let o = o.into_object().map_err(|e| anyhow!("{e}"))?;
+            let Some(SuiRawData::Package(p)) = o.bcs else {
+                bail!("Expected SuiRawData::Package but got something else");
+            };
+            p.to_move_package(u64::MAX /* safe as this pkg comes from the network */)
+                .map_err(|e| anyhow!("{e}"))
+        })
+        .collect()
+}
+
+/// Filter out a package's dependencies which are not referenced in the source code.
+pub(crate) async fn pkg_tree_shake(
+    client: &SuiClient,
+    with_unpublished_dependencies: bool,
+    compiled_package: &mut CompiledPackage,
+) -> Result<(), anyhow::Error> {
+    let pkgs = compiled_package.find_immediate_deps_pkgs_to_keep(with_unpublished_dependencies)?;
+    let pkg_ids: Vec<_> = pkgs.clone().into_values().collect();
```

Comment:
Cloning a larger data structure to select some part of it is a bit of an anti-pattern -- take the part and just clone that.

```suggestion
    let pkg_ids: Vec<_> = pkgs.values().cloned().collect();
```